Feat nrf ticks support

hoihu opened the pull request:
This is a follow up / replacement of https://github.com/micropython/micropython/pull/5470

It supports `time.ticks_ms(x)`, `time.ticks_add(x)` and `time.ticks_diff(x)` for two configurable timebases:

1) Ticker timebase (1MHz), precise, app 5uA, 32bits
2) RTC timebase (32kHz), less precise, app 0.5uA, 24bits

both are capabale of running in WFI low power mode (in contrast to the SYSTICK module)
see above link for further discussions / insights / thoughts).

hoihu committed:
port/nrf: set ticker's CCR0 timebase to 1usec.

hoihu committed:
port/nrf: move rtcounter's clock init to init0.

hoihu committed:
ports/nrf: add tick support for time module.

The time tick support is enabled with the MICROPY_PY_TIME_TICKS
compiler switch.

By default it uses the Ticker's 1MHz clock to schedule a CCR0
callback in 1msec intervals (MICROPY_PY_TIME_USE_TICKER_BASE)
Using the 1MHz clock typically consumes app. 5uA.
It's a 32bit counter and overflows every 50 days

For very power sensitive applications the RTC can also be used
as timebase using the MICROPY_PY_TIME_USE_RTC_BASE compiler switch.
In this mode, an error of app 0.7% is introduced due to the granularity
of the 32kHz ticks vs 1msec ticks. The RTC clock consumes app. 0.5uA.
Since it's a 24bit counter, overflows occurs every 4h.

Both modes are operational in WFI power save mode, in contrast
to the SYSTICK module which switches off.

hoihu commented on ports/nrf/mpconfigport.h:
```diff
@@ -296,6 +307,11 @@ extern const struct _mp_obj_module_t ble_module;
 
+#ifdef MICROPY_PY_TIME_USE_RTC_BASE
+// configure 24 bit time base for ticks
+#define MICROPY_PY_UTIME_TICKS_PERIOD (1 << 24)
```
@dpgeorge : could you please verify that I got the  define right for a 24bit timebase?

dpgeorge replied:
It looks correct, and it should be easy to test because 2**24 microseconds is only 16 seconds.  So if this were set to `1<<25` then ticks diff (of microseconds) will fail after 16 seconds.  I will test this myself to double check.

dpgeorge replied:
Except I just realised there is no `ticks_us()`.... is it possible to make one using the result of `NRF_TIMER1->COUNTER`?  Of course that will only work for `MICROPY_PY_TIME_USE_TICKER_BASE`.

hoihu replied:
> It looks correct, and it should be easy to test because 2**24 microseconds is only 16 seconds. So if this were set to 1<<25 then ticks diff (of microseconds) will fail after 16 seconds.

The timebase for the RTC is 1.0071msec using the HW prescaler, so the overflow occurs after 4.7 hours. 

Shouldn't be an issue anymore with the proposal below though: the overflows with 32kHz tick rate @ 24bits occurs every 512 sec and that would be handled by the irq. 

The actual msec value calculated there would then be given back to tick_msec(), and that's likely a 32bit value.

hoihu replied:
> is it possible to make one using the result of NRF_TIMER1->COUNTER

As I understand the Nrf needs a CCRx register to read the counter value. See my comment https://github.com/micropython/micropython/pull/5470#issuecomment-587747871 on how I did this. I believe only the RTC's counter value can be accessed directly.

I didn't include this because I wanted to come through with the standard ticks_msec first ;) But I can go for it too.. Do we want to spend CCR1 for the usec timebase?

tve replied:
I don't know anything about the Nrf port, so this may be a non-issue, but please keep in mind that a real-time clock is different from a CPU cycle counter clock. Real-time clocks can get synchronized to the real world and therefore stretched or stepped. ticks/ticks_ms/ticks_us shouldn't be affected by that.

WayneKeenan replied:
@tve fyi: RTC in this context stands for Real Time Counter, it's not a real world clock.

dpgeorge replied:
> I didn't include this because I wanted to come through with the standard ticks_msec first

You're right, that's definitely a good goal, to just get `ticks_ms` working first.

hoihu replied:
outdated with better solution, see below

hoihu committed:
ports/nrf: enable time tick support.

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,57 @@
+}
+
+void ticker0_init_msec(void) {
+    set_ticker_callback(0, tick_cb, 200);
```
Why 200 as the initial callback?  Could just make it 0, or 1000?

hoihu replied:
that's an arbitrary value. Yes could be 0 or 1000. Will set it to 1000.

hoihu replied:
outdated by using RTC

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,57 @@
+}
+
+mp_uint_t mp_hal_ticks_ms(void) {
+    return (mp_uint_t)rtc1.p_reg->COUNTER;
```
To make this more accurate, would it be possible to set the prescaler to 0, so the RTC counts at 32kHz, then return `COUNTER * 1000 / 32768`?  It would also need to keep track of roll-over of the RTC counter to get longer times, which could be done in an IRQ if the RTC has an IRQ?

hoihu replied:
Good idea. 
It definetly helps with the long-term accuracy of the time, something that the user should except when running from the 32kHz crystal. It might not help for short time intervals (e.g. if you want to delay for 3msec). But for such short intervals, the user might need to use `time.usec` anyway.

> if the RTC has an IRQ?

yes it has one, and the overflow event can be set as irq.

I'll check this if I can come up with something.

dpgeorge replied:
Sounds good.  It will actually have good ms-resolution, because the underling RTC counter has sub-ms resolution.

hoihu commented:
And thinking on this a bit further... 

With the proposal of using the RTC as by @dpgeorge , the advantage of using the Ticker's timebase is almost neglectable IMO. Actually the RTC will be more precise for longer intervals this way than the ticker's 1MHz because it is sourced by the high precision 32kHz crystal (if it is present..). 

So I wonder if we should use CCR0 of the ticker for the usec timebase, and make the RTC the default and only implementation for the msec timebase? 

Or we invest CCR1 to support the usec timebase and leave the rest configurable as is (RTC or ticker based msec timebase).

hoihu committed:
ports/nrf: use RTC clock more efficiently

this approach uses a prescaler of 0 and handles the overflow
event in an irq

hoihu commented:
Ok so I've pushed an update on how it could look with RTC overflow bit handling. It doesn't add much code.

It works on my PCA10040 devkit (running more than 512 sec didn't overflow).

dpgeorge commented:
> Actually the RTC will be more precise for longer intervals this way than the ticker's 1MHz because it is sourced by the high precision 32kHz crystal (if it is present..).
> 
> So I wonder if we should use CCR0 of the ticker for the usec timebase, and make the RTC the default and only implementation for the msec timebase?

I think this is a good approach, to have RTC1 as the only implementation of msec.

And in fact it can also be used for usec, just with a ~30us resolution.  Simply return `RTC1->COUNTER * 1000000 / 32768`.  For low-power applications I think 30us resolution is pretty good considering it has minimal overhead, simple implementation, and no additional resource usage.

In summary so far, if only for my own understanding of the situation:
- the nrf port doesn't have `ticks_ms` or `ticks_us`
- SysTick could be used to implement these in the same way it's done on stm32, except that SysTick on nrf doesn't run during a WFI
- instead RTC is a good choice, it's essentially a low-power replacement for SysTick which runs during sleep
- RTC0 is used by soft-device, not all MCUs have RTC2, so use RTC1 (dedicate it to internal timing, not for user use)
- set `RTC1->PRESCALER=0` to get maximum resolution in `COUNTER`, implement overflow to get more than 24-bit precision, and use this to construct a millisecond counter for `ticks_ms`
- can also construct a microsecond counter for `ticks_us` with 30us resolution, at no extra cost
- could use ticker's CCR0/1 for a 1us-resolution `ticks_us`, at extra power cost and usage of ticker peripheral (IMO this is not worth doing at this stage)
- could eventually use RTC1's CCRs to implement an efficient softtimer for `machine.Timer()`, following stm32's approach (but that's out of scope of this PR)

hoihu commented:
yes that is the situation from my point of view too.

I‘ll  cleanup the PR accordingly: add the usec support using RTC, remove the tickers msec support. 

should there be a check in the rtccounter code that will prevent using RTC1 when it is used by the time module, raising a error?

dpgeorge commented:
> should there be a check in the rtccounter code that will prevent using RTC1 when it is used by the time module, raising a error?

On the one hand it might be useful to let the user inspect RTC1 and maybe configure it if they know what they're doing (even when it's used by the time module).  But OTOH it will be very confusing for those that don't know RTC1 is used by the time module, when they change it and all the time functions start behaving strangely.  So I'd say raise an exception.

glennrub commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,76 @@
+    .reliable     = 0,
+    .tick_latency = 0,
+    #ifdef NRF51
+    .interrupt_priority = 3,
```
Could this be to set priority 1 for nrf51 and priority 3 for others? 

I'm not having a very good example, but imagine someone using machine.RTC/machine.Timer, to issue a timed operation like poll(timeout) on regular intervals. As the callback is in IRQ priority 3, and RTC1 is in priority 3, these will chain, and the overflow event might not get through (given that mp_hal_ticks_ms() is the source of the timed operation). 

The priorities proposed should be fine with SoftDevices, as these are free for application use. And the fact that the overflow routine is this small, it should not impact any performance of the stacks.

hoihu replied:
> Could this be to set priority 1 for nrf51 and priority 3 for others?

I have no strong opinion on this. If you think that's better I have no problem adapting it.

> As the callback is in IRQ priority 3, and RTC1 is in priority 3, these will chain, and the overflow event might not get through (given that mp_hal_ticks_ms() is the source of the timed operation).

I think you are relating to the time slot where the overflow has occured but something prevents the rtc irq handler being called (e.g. by a already running irq)? During this time,  `mp_hal_ticks_ms()` would return an invalid value since the counter's value is 0 and the overflow counter hasn't increased. Assuming that `mp_hal_ticks_ms()` is not called in a irq context then it's not a problem but otherwise it could create a problem (very rare though, every 512 sec for a very short time and assuming that `mp_hal_ticks_ms()` is called in irq). 

Otherwise, `mp_hal_ticks_ms()` could be guarded by a check to the overflow flag... Perhaps that's the simplest solution to this.

Just wondering if nested irq's enabled?

glennrub replied:
I agree, its not a problem if `mp_hal_ticks_ms` is not called from IRQ context. And this is a very very tiny time window to hit it, but it might lead to a deadlock.

> Just wondering if nested irq's enabled?

IRQ's for the peripherals are nested. The nrfx drivers use the `nrfx_glue.h` in the nrf-port to map the NVIC functions.

> Otherwise, mp_hal_ticks_ms() could be guarded by a check to the overflow flag... Perhaps that's the simplest solution to this. 

I like the idea :) it would make my concern about this "IRQ in IRQ" go away. Letting `mp_hal_ticks_ms()` read and clear the OVERFLOW event flag (bypassing the nrfx driver enabling the interrupt, and manually enable the event). But, how would we enforce `mp_hal_ticks_ms()` to be run at least once between overflows? IMO keeping the IRQ handler to detect overflow as you have it already is better after all, and raise the IRQ priority.

hoihu replied:
> Letting mp_hal_ticks_ms() read and clear the OVERFLOW event flag

Yeah I was just digging through this and to make things simple: why not, on OVERFLOW detected,  just return the last valid tick value and make the irq clearing the event bit? 

We would need to store the last valid value though.

> But, how would we enforce mp_hal_ticks_ms() to be run at least once between overflows?

Ah we may have misunderstood each other. My idea was to leave the irq handler as is and make the guarding only to check if an overflow is pending. So there is no need to enforce a call.

hoihu replied:
> IRQ's for the peripherals are nested.

hmm that makes it worse.. then even `rtc_irq_time`can be interrupted (possibly even in the nrfx HAL code part..) if that happens and the higher irq handler uses the tick method, this would lead to the same thing as before. 

I mean we could agree on not using the `mp_hal_ticks_ms()` within irqs?

glennrub replied:
> then even rtc_irq_timecan be interrupted (possibly even in the nrfx HAL code part..)

Do you mean the other peripherals? or the RTC?

> I mean we could agree on not using the mp_hal_ticks_ms() within irqs?

I know it is not the best code pattern, but it will be code that does it. It might be hard to not do it when `machine.Timer` and `machine.RTCounter` has callbacks to fill. These are triggered from IRQ context in its given priority.

Or, maybe agree that we only call `mp_hal_ticks_ms()` from a lower IRQ or main thread? :)
This is the case for most of the machine modules. They are configured to priority 3 and 6.

dpgeorge replied:
For `mp_hal_ticks_us()` (note us not ms) stm32 handles overflow both in the SysTick IRQ and in the call to `mp_hal_ticks_us()`.  And SysTick IRQ runs at the highest priority.

> Or, maybe agree that we only call `mp_hal_ticks_ms()` from a lower IRQ or main thread?

That might be a good solution.

hoihu replied:
> Do you mean the other peripherals? or the RTC?

I meant that the RTC's irq entry point is in the nrfx lib and is eventually dispatched to the subscribed handler(s). So `rtc_irq_time` is really just the handler part. The majority of the irq time is probably spend in the nrfx part, checking and calling the handlers etc. And during this time, the irq can be preempted also.

> Or, maybe agree that we only call mp_hal_ticks_ms() from a lower IRQ or main thread? :)
This is the case for most of the machine modules. They are configured to priority 3 and 6.

I also think that's a good proposal.

hoihu replied:
>  Or, maybe agree that we only call mp_hal_ticks_ms() from a lower IRQ or main thread? :)

sigh sorry for coming back to the subject, but I think this doesn't work either since `mp_hal_ticks_ms` can be preempted by the RTC irq. Specifically the line `return (rtc_overflows << 9) * 1000 + ((mp_uint_t)rtc1.p_reg->COUNTER * 125 / 4096);` can interrupted by the RTC irq. If that happens, the RTC counter is at 0 but the rtc_overflows counter had the old value so the caluclated value is actually wrong. It's a tiny time window, but still possible to hit. 

What could be done is to guard the calculation against the irq of the RTC like: 

- disable RTC irq. 
- make the calculation. 
- reenable RTC irq again. 
- return tick value.

hoihu replied:
Ok I've pushed a version that does the above guarding. Let me know if the current version is so far ok. Then I'll cleanup the commit mess ;)

hoihu replied:
see below comments for a better solution

glennrub commented:
@hoihu, As i mentioned in the earlier PR, i was hunting for a timer replacement for nrf91 sockets. I have tested the current code, and it works very well :) Thanks a lot for doing this!

hoihu committed:
ports/nrf: add time.ticks_us and cleanup of of code.

hoihu commented:
So this is the work in progres so far, not yet cleaned up and not handling the overflow condition within `mp_hal_ticks_ms()`. Couple of things I've pointed out, in `mphalconfigport.c` as comments.

glennrub commented on ports/nrf/mpconfigport.h:
```diff
@@ -163,7 +163,7 @@
 
 #ifndef MICROPY_PY_MACHINE_SOFT_PWM
-#define MICROPY_PY_MACHINE_SOFT_PWM (0)
+#define MICROPY_PY_MACHINE_SOFT_PWM (1)
```
Are there any dependencies left to `SOFT_PWM`? Could this be turned off by default again?

hoihu replied:
Sure

dpgeorge commented on ports/nrf/drivers/ticker.c:
```diff
@@ -102,9 +102,10 @@ static ticker_callback_ptr callbacks[3] = { noop, noop, noop };
     if (ticker->EVENTS_COMPARE[0]) {
         ticker->EVENTS_COMPARE[0] = 0;
-        ticker->CC[0] += call[0]()*MICROSECONDS_PER_TICK;
+        ticker->CC[0] += call[0]();
```
Is this change needed anymore?

hoihu replied:
No, I'll cleanup that

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,71 @@
+    // An easier solution is to use COUNTER * 32, but introduces an error
+    // of app. 5%. However lets consider this as a tradeoff since the tick's rate
+    // is also very rough and not really suitable for low value usec timings.
+    return (mp_uint_t)rtc1.p_reg->COUNTER * 32;
```
It can use `COUNTER * 214 / 7` which gives an error of 0.18% and will not overflow 32 bits.

hoihu replied:
Ok -  I was thinking about a simple division using a 2**x divisor to minimize overhead. But actually that's not really needed. Will adapt this.

dpgeorge replied:
I understand.  But I think the overhead of `/7` is small enough (the `*214` should be very efficient even without hardware mult).

hoihu replied:
agree

hoihu committed:
ports/nrf: use optimized formula for usec calculation

also revert ticker's changes and SOFT_PWM settings
guard mp_hal_ticks_ms() against preempting of RTC overflow irq

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,82 @@
+    //
+    // An easier solution is to use COUNTER * 214 / 7, introducing a small
+    // error of app. 0.18%.
+    return (mp_uint_t)rtc1.p_reg->COUNTER * 214 / 7;
```
Thinking about this more, this won't wrap correctly at 0xffffffff, it'll wrap after its maximum value of 0x1e924906.   It also needs to do a power-of-2 division to make the overflow work correctly (so you were right to have it a simple *32 before).  (Eg `0xffffffff * 214/7 = 0x1e92492473` which, if it overflows to `0x0 * 214/7`, will give a big jump in ticks; and no amount of precision will make this work.)

So it needs similar logic to `mp_hal_ticks_ms()` and a 64-bit mult/divide.  This can be done explicitly without much cost like this:
```c
// same code as mp_hal_ticks_ms() to get overflows and counter in an atomic way
// ...

// compute: ticks_us = (overflows << 24 + counter) * 1000000 / 32768
//    = (overflows << 15 * 15625) + (counter * 15625 / 512)

// first compute counter * 15625
uint32_t counter_lo = (counter & 0xffff) * 15625;
uint32_t counter_hi = (counter >> 16) * 15625;
// actual value is counter_hi << 16 + counter_lo

uint32_t ticks_us = ((overflows << 15) * 15625) + ((counter_hi << 7) + (counter_lo >> 9));
```
I tested the above by hand at some boundary values, but you may want to double check it.

hoihu replied:
this works!

glennrub commented:
> sigh sorry for coming back to the subject, but I think this doesn't work either since mp_hal_ticks_ms can be preempted by the RTC irq.

Could it be an option to handle counter read in IRQ context as well, a bit like it is done in [machine/rtcounter.c](https://github.com/micropython/micropython/blob/master/ports/nrf/modules/machine/rtcounter.c#L91)? Even if not exactly the same implementation, you can get both the Compare match event and Overflow event in the same nrfx-hal interrupt loop.  Then use the counter value and overflow information to update a shared volatile variable with `mp_hal_ticks_ms`.

hoihu commented on ports/nrf/modules/machine/rtcounter.c:
```diff
@@ -174,11 +186,6 @@ STATIC mp_obj_t machine_rtc_make_new(const mp_obj_type_t *type, size_t n_args, s
 
-    // Start the low-frequency clock (if it hasn't been started already)
-    if (!nrf_clock_lf_is_running(NRF_CLOCK)) {
-        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
```
ok

hoihu replied:
done

hoihu commented on ports/nrf/modules/machine/rtcounter.c:
```diff
@@ -108,6 +109,10 @@ STATIC void interrupt_handler2(nrfx_rtc_int_type_t int_type) {
 void rtc_init0(void) {
+    // Start the low-frequency clock (if it hasn't been started already)
+    if (!nrf_clock_lf_is_running(NRF_CLOCK)) {
+        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
```
ok

hoihu replied:
done

hoihu commented on ports/nrf/main.c:
```diff
@@ -226,6 +227,10 @@ int main(int argc, char **argv) {
 
+    #if MICROPY_PY_TIME_TICKS
+    rtc1_init_time_ticks();
+    #endif
```
ok

hoihu replied:
done

hoihu commented:
> Could it be an option to handle counter read in IRQ context as well, a bit like it is done in machine/rtcounter.c?

Maybe.. But how can we trigger the irq (e.g. from mp_hal_ticks_ms())? Is there a way of manually fire a RTC irq? And there are still race conditions left I believe.. since the overflow event is still an IRQ & If at the same time mp_hal_ticks_ms() has fired it's irq requests they may conflict.

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,82 @@
+    // overflow irq while the calculation takes place
+    mp_uint_t ticks;
+    rtc1.p_reg->INTENCLR = RTC_INTENSET_OVRFLW_Msk;
+    ticks = (rtc_overflows << 9) * 1000 + ((mp_uint_t)rtc1.p_reg->COUNTER * 125 / 4096);
```
> Hmm.. why is the check for counter necessary? If overflow is set it can be safely assumed that overflows must be increased by 1?

If OVRFLW is set it could be before or after COUNTER is read.  If before then the adjustment must be made, if after then no adjustment.  The before case is when COUNTER is very small (because it just overflowed and was set to zero), the after case is when COUNTER is very large (because it's just about to overflow but we read it right before it overflows).  The extra check for counter is to distinguish these cases and I chose `1<<23` because it's halfway between min and max values of COUNTER.

> Also before enabling the irq we should reset the irq flag since the rtc_irq is otherwise immediately called & this increases the overflow counter again...



No.  The global `rtc_overflows` still needs to be incremented.  Only the local copy here is adjusted by 1 if needed.

> Also to simplify things.. Perhaps it just make sense to combine the code altogether and calculate a 64bit wide usec tick counter for both ticks_ms and ticks_us as you pointed out below. For `mp_hal_ticks_ms()` we'd just in addition divide by 1000 & mask by 0xffff ffff.

But a div-by-1000 of a uint64_t will be expensive.  I think it's ok to duplicate code like this, it's time critical and needs to be precise.

> This complexity makes me think if we shouldn't just reconsider and move to the timer's 32bit 1MHz clock?

If we want both ms and us then there needs to be some complexity somewhere, unless there are two 32-bit dedicated timers, one running at 1kHz for ms and one at 1MHz for us.

hoihu replied:
thanks for the clarification, now I understand. Agree with argumentation. Will look into this (it's kind of getting really interesting here ;))

hoihu replied:
done. I copy&pasted your comment above to the code if that's ok for you. I believe it clarifies the issue nicely.

hoihu committed:
ports/nrf: move rtc init at beginning of main

hoihu committed:
ports/nrf: revert changes to clock init.

hoihu committed:
ports/nrf: change calculation of msec, usec

and add MICROPY_EVENT_POLL_HOOK support by setting rtc to app fire 1msec irqs
rewrite mp_hal_delay_us to use ticks_us and mp_hal_delay_ms to use ticks_ms
mp_hal_delay_ms is using MICROPY_EVENT_POLL_HOOK

hoihu commented:
Ok thanks @dpgeorge and @glennrub for all your inputs! 

I've added in addition the changes to `mp_hal_delay_ms` and `mp_hal_delay_us`. `mp_hal_delay_ms` now introduces `MICROPY_EVENT_POLL_HOOK` support with WFI. In order to leave WFI I've setup a irq for CC1 that gets called every 1msec (and not handled, but it wakes up the CPU). This is then the foundation on which `select.poll`and asyncio should work. I think it makes a lot of sense to introduce these changes in the PR too, but let me know your opinion about it.

dpgeorge commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,120 @@
+        overflows += 1;
+    }
+    rtc1.p_reg->INTENSET = RTC_INTENSET_OVRFLW_Msk;
+    // enable_irq(state);
```
would be good to factor this chunk out to a macro (eg `RTC1_GET_TICKS_ATOMIC`) and use it here and above in `mp_hal_ticks_ms()`

hoihu replied:
ok done

dpgeorge commented:
> `mp_hal_delay_ms` now introduces `MICROPY_EVENT_POLL_HOOK` support with WFI. In order to leave WFI I've setup a irq for CC1 that gets called every 1msec (and not handled, but it wakes up the CPU). This is then the foundation on which `select.poll`and asyncio should work.

That's OK for now.  Eventually the plan is to move to something more event based so there's not a constant polling (see #6110) but for now a WFI with wake-up is OK.

glennrub commented:
> CC1 that gets called every 1msec (and not handled, but it wakes up the CPU).

You need to handle the compare0 event, to set a new CC. I'll point it out in the code. When doing that, the interval measures 0.9276 ms on logic analyzer.

glennrub commented on ports/nrf/mphalport.c:
```diff
@@ -35,6 +35,120 @@
+    // irq handler for overflow
+    if (event == NRFX_RTC_INT_OVERFLOW) {
+        rtc_overflows += 1;
+    }
```
If you add the following snippet, the CC will be periodic: 
```
if (event == NRFX_RTC_INT_COMPARE0) {
        uint32_t new_cc = nrfx_rtc_counter_get(&rtc1) + 30;
        nrfx_rtc_cc_set(&rtc1, 0, new_cc, true);
}
```

hoihu replied:
Yes I did that just 30min ago, sorry ;)

hoihu replied:
and 33 as increase to be a bit more precise with 1msec.

hoihu committed:
nrf/ports: use macros for generic rtc irq handling.

hoihu committed:
ports/nrf: use old delay functions when ticks support is not enabled

hoihu commented:
pushed an update
- use macros for common "guarding" code of RTC overflow (as by review @dpgeorge)
- fixed a bug with the CCR handling (now it correctly uses CCR0 and ~1msec intervals for wakeup)
- reintroduce legacy delay functions if tick support is not enabled
- some minor cleanup

I tested this always with the PCA10040 btw.
If that looks ok, I'll start cleanup the commits. Looking forward for feedback!

hoihu commented:
I've placed https://github.com/micropython/micropython/pull/6202 for a cleaned up version.

tannewt commented:
FYI, I switched CircuitPython to the RTC (from SysTick) on our supported ports as well. In cases where we need higher precision (such as PulseIn) we use a separate timer.

hoihu commented:
> FYI, I switched CircuitPython to the RTC (from SysTick) on our supported ports as well. In cases where we need higher precision (such as PulseIn) we use a separate timer.

Good to know, thanks for the information.

dpgeorge commented:
I'll close this because it's superseded by the implementation in 15574cd665a8e66157aa97447e3869513da81d45
